File: rb/test/competition_entity_test.rb
# Competition entity test

require "minitest/autorun"
require "json"
require_relative "../WorldCupQualification_sdk"
require_relative "runner"

class CompetitionEntityTest < Minitest::Test
  def test_create_instance
    testsdk = WorldCupQualificationSDK.test(nil, nil)
    ent = testsdk.Competition(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = competition_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "competition." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set WORLDCUPQUALIFICATION_TEST_COMPETITION_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    competition_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.competition")))
    competition_ref01_data = nil
    if competition_ref01_data_raw.length > 0
      competition_ref01_data = Helpers.to_map(competition_ref01_data_raw[0][1])
    end

    # LIST
    competition_ref01_ent = client.Competition(nil)
    competition_ref01_match = {}

    competition_ref01_list_result, err = competition_ref01_ent.list(competition_ref01_match, nil)
    assert_nil err
    assert competition_ref01_list_result.is_a?(Array)

    # LOAD
    competition_ref01_match_dt0 = {
      "id" => competition_ref01_data["id"],
    }
    competition_ref01_data_dt0_loaded, err = competition_ref01_ent.load(competition_ref01_match_dt0, nil)
    assert_nil err
    competition_ref01_data_dt0_load_result = Helpers.to_map(competition_ref01_data_dt0_loaded)
    assert !competition_ref01_data_dt0_load_result.nil?
    assert_equal competition_ref01_data_dt0_load_result["id"], competition_ref01_data["id"]

  end
end

def competition_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "competition", "CompetitionTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = WorldCupQualificationSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["competition01", "competition02", "competition03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["WORLDCUPQUALIFICATION_TEST_COMPETITION_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "WORLDCUPQUALIFICATION_TEST_COMPETITION_ENTID" => idmap,
    "WORLDCUPQUALIFICATION_TEST_LIVE" => "FALSE",
    "WORLDCUPQUALIFICATION_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["WORLDCUPQUALIFICATION_TEST_COMPETITION_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["WORLDCUPQUALIFICATION_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = WorldCupQualificationSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["WORLDCUPQUALIFICATION_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["WORLDCUPQUALIFICATION_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/competition_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/world-cup-qualification-sdk/go"
	"github.com/voxgig-sdk/world-cup-qualification-sdk/go/core"

	vs "github.com/voxgig-sdk/world-cup-qualification-sdk/go/utility/struct"
)

func TestCompetitionEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Competition(nil)
		if ent == nil {
			t.Fatal("expected non-nil CompetitionEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"competition": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Competition(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Competition(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := competitionBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "competition." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WORLD_CUP_QUALIFICATION_TEST_COMPETITION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		competitionRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.competition", setup.data)))
		var competitionRef01Data map[string]any
		if len(competitionRef01DataRaw) > 0 {
			competitionRef01Data = core.ToMapAny(competitionRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = competitionRef01Data

		// LIST
		competitionRef01Ent := client.Competition(nil)
		competitionRef01Match := map[string]any{}

		competitionRef01ListResult, err := competitionRef01Ent.List(competitionRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, competitionRef01ListOk := competitionRef01ListResult.([]any)
		if !competitionRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", competitionRef01ListResult)
		}

		// LOAD
		competitionRef01MatchDt0 := map[string]any{
			"id": competitionRef01Data["id"],
		}
		competitionRef01DataDt0Loaded, err := competitionRef01Ent.Load(competitionRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		competitionRef01DataDt0LoadResult := core.ToMapAny(entityData(competitionRef01DataDt0Loaded))
		if competitionRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if competitionRef01DataDt0LoadResult["id"] != competitionRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func competitionBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "competition", "CompetitionTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read competition test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse competition test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"competition01", "competition02", "competition03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WORLD_CUP_QUALIFICATION_TEST_COMPETITION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WORLD_CUP_QUALIFICATION_TEST_COMPETITION_ENTID": idmap,
		"WORLD_CUP_QUALIFICATION_TEST_LIVE":      "FALSE",
		"WORLD_CUP_QUALIFICATION_TEST_EXPLAIN":   "FALSE",
		"WORLD_CUP_QUALIFICATION_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["WORLD_CUP_QUALIFICATION_TEST_COMPETITION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WORLD_CUP_QUALIFICATION_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["WORLD_CUP_QUALIFICATION_APIKEY"],
			},
			extra,
		})
		client = sdk.NewWorldCupQualificationSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WORLD_CUP_QUALIFICATION_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WORLD_CUP_QUALIFICATION_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
